### PR TITLE
[wip] Implement webhooks API

### PR DESCRIPTION
**Pull request by lappemic**
Hey @Wauplin, here's the initial draft for #1808, trying to align with #1905 as you suggested. However, a couple of points are still unclear to me, and I'd love to get your input to ensure the implementation aligns correctly:
- Could you clarify what the `secret` parameter represents in `create_webhook` and `update_webhook` methods? Is it intended to be the user token?
- Regarding the definition of `DOMAIN_T` Literal, would it be appropriate to define it directly within the function signatures?

I plan to finalize docstrings, module imports, and tests once we're on the same page regarding these questions.

Thanks a lot for your guidance and feedback!

### CODE REVIEW

**Comment by Wauplin** [comment]
Thanks @lappemic!

I didn't have a look yet to your changes but to answer your question:
- the secret correspond to a value set by the user and that will be sent in every webhook. This way the user is sure the webhook is sent by the Hub. So this is not the same as `token` no. You can have more info here: https://huggingface.co/docs/hub/webhooks
- if it's used only once, then you don't need to define `DOMAIN_T` explicitly. If used more than once, yes it's best to define it (maybe called `WEBHOOK_DOMAIN_T`). 

Also 2 comments about the changes:
- it's best to define a dataclass rather than a typed dict as return value
- there are a lot of changes due to styling. Please run `pip install -e '.[dev]'` from your local folder to install all the required deps and then run `make style` to comply with huggingface_hub standard. You can run `make quality` to check everything's fine.

**Comment by Wauplin** [comment]
> Since WEBHOOK_DOMAIN_T is utilized more than once, I'll go ahead and define it in constants.py. Does that sound right?

Yes!

> I would think it is about the WatchedItem definition to do rather like this

Exactly! That will be easier to use for users downstream

> Regarding the style changes, I had previously run make style and make quality thinking they'd align my code with Hugging Face's style guidelines. However, it seems some styling issues persist. I've run both commands again without any changes. Any clue what might be causing this?

Hmm that's weird. I would assume something's off with the install. Are you using a virtualenv? If no, I'd advice to create a new one (installing Python3.9 or 3.10 in it) and install the dependencies in it. You can check install guidelines here: https://huggingface.co/docs/huggingface_hub/installation#install-with-pip

**Comment by lappemic** [comment]
Thank you for the feedback!

I've updated the PR with the latest commits, according to your suggestions.

Regarding the styling issues, I initially followed the steps outlined in the repo's [CONTRIBUTING.md](https://github.com/huggingface/huggingface_hub/blob/main/CONTRIBUTING.md) without success. Then, I set up a new virtual environment and followed the installation instructions from [https://huggingface.co/docs/huggingface_hub/installation#install-with-pip](https://huggingface.co/docs/huggingface_hub/installation#install-with-pip), which worked now. The only difference is the `[dev]' notation. I assume this should not make any difference?

Looking forward to your review and approval to continue.











Thanks a lot for the inputs!

The latest commits incorporate your suggestions. I am going to wait for your approval to go on for the further implementation.

A point about the styling: before i followed the instructions in this repos [CONTRIBUTING.md](https://github.com/huggingface/huggingface_hub/blob/main/CONTRIBUTING.md), which failed. Afterwards i created a new venv according to this source: [https://huggingface.co/docs/huggingface_hub/installation#install-with-pip](https://huggingface.co/docs/huggingface_hub/installation#install-with-pip)

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8531]: It's best to duplicate the `args` section here. This is extra maintenance work but very helpful for IDEs and to correctly generate the documentation.

**Review by Wauplin — COMMENTED** [review]
Thanks @lappemic! I'll have a closer look to it when I have time later this week. It seems that the styling still have some issues (see [here](https://github.com/huggingface/huggingface_hub/pull/2209/files#diff-71a3e23532e1355e79edd51b351bd514f0aa55b946e4c530b0c95f647773ac5dR107)) but I can take care of that if you want. Thanks again!

File: src/huggingface_hub/hf_api.py
Comment (lappemic) [reply src/huggingface_hub/hf_api.py]: I've updated the docstring as suggested in the upcoming commit.

I'll proceed with the remaining updates once you've reviewed the implementation.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:395]:
```suggestion
class WebhookWatchedItem:
    """Data structure containing information about the items watched by a webhook.
```

I'd rename it `WebhookWatchedItem` to avoid confusion with other dataclasses unrelated to webhooks. This change has to be reflected in every occurrence of `WatchedItem`

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8286]: Missing args + return type in docstring :)

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8296]: Same here

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8285]:
```suggestion
    def get_webhook(self, webhook_id: str, token: Optional[str] = None) -> Webhook:
```

=> would be good to type the output value with a dataclass (called `Webhook`?). This way users will know which fields to expect and IDEs will support autocompletion.

This meant the `return response.json()` must be adapted as well.

Comment (Wauplin) [reply src/huggingface_hub/hf_api.py:8285]: Same comment for `list_webhooks`, `create_webhook`, `update_webhook`

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8318]:
```suggestion
            watched (`List[WatchedItem]`):
                List of [`WatchedItem`] to be watched by the webhook. It an be users, orgs, models, datasets or spaces.
```

2 things:
- better to have backquotes when defining the type
- setting ```[`WatchedItem`]``` in a docstring will automatically link to the object section in the docs. This is very convenient for users reading the docs on https://huggingface.co/docs/huggingface_hub/index

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8329]:
```suggestion
            [`Webhook`]: The created webhook.
```

(nit) same for returns value that can be referenced (once `Webhook` dataclass will be defined)

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8333]: Thanks for adding a proper example! :heart:

**Review by Wauplin — COMMENTED** [review]
Thanks for the PR @lappemic! I just reviewed it and made high level comment to it. The structure is good and I think you can continue on that. Sometimes I commented on a specific method but the comment can be applied to all comment.

As you can see, I also pushed commits to fix styling stuff. Make sure not to push changes that are unrelated to your PR, otherwise it makes the review very difficult (I have to "check" if a change is related to your PR or to some styling). If you correctly install the `[dev]` dependencies + run `make style`, it shouldn't happen. Anyway, let me know if you have questions related to this. And don't forget to `git pull` before starting to make changes.

**Comment by lappemic** [comment]
Thanks you very much for the proper review and inputs @Wauplin. I hope to find time to incorporate everything this week. It might take up to next week though.

**Comment by Wauplin** [comment]
Thanks @lappemic! And no worries about the timing, days are always too busy :smile:

**Comment by lappemic** [comment]
Hey @Wauplin, found some time to implement your feedback.

I was not quite sure if `enable_webhook` and `disable_webhook` also should return the `Webhook`? I assumed yes, so the users gets instant feedback.

As soon as the main changes are alright, i will implement the tests and extend the documentation accordingly.

File: src/huggingface_hub/__init__.py
Comment (Wauplin) [inline src/huggingface_hub/__init__.py:169]: I'd also add `Webhook` object (or rather `WebhookInfo`?) to `__init__.py`.

**Review by Wauplin — COMMENTED** [review]
Very nice PR @lappemic! It is a pleasure to review such a clean PR :)

I've left minor comments, otherwise everything looks good to me :+1: Do you have questions regarding tests or it is all fine?

Regarding the return value of `enable_webhook`/`disable_webhook`, if the server returns a value, then we must forward this value to the user. As a client library, `huggingface_hub` shouldn't hide something returned by the server -if any-.
If the server don't return anything (I don't remember TBH), then it's fine to return None.

File: src/huggingface_hub/hf_api.py
Comment (lappemic) [reply src/huggingface_hub/hf_api.py:8437]: Probably i am overdoing it with references :smile:

I was confused with the referencing style: Is it always enough to reference just like [`WebhookInfo`] then? If yes, why are there examples of [`huggingface_hub.hf_api.RepoSibling`] or [`hf_api.DatasetInfo`]? Is there a need of standardising this as well? Things like this increases in my opinion fraction for contributors to easily decide which way to go (only if they do not know it anyway of course :D).

**Comment by lappemic** [comment]
Thank you so much @Wauplin 🙏 You really know how to motivate contributors 😅 

Your comments are all included in the last commits. 👍 

Regarding tests: I would assume i write them in `tests/test_hf_api.py`? Except from that i would just give it a try and see how far it gets me. :D If this is ok for you? I assume i would just need some inputs at one point. Except you know already something which might be a blocker for someone with my skillset atm. 😄 

About `enable_webhook` and `disable_webhook`: How do i find out if the server returns something? 😬

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [reply src/huggingface_hub/hf_api.py]: What I meant is that in
```py
        Returns:
            [`Webhook`]:
                The created [`huggingface_hub.hf_api.Webhook`] configuration.
```

there is no need to reference twice the `Webhook` object (though it's valid).
Regarding the format,  
```
[`Webhook`]
```
and
```
[`huggingface_hub.hf_api.Webhook`]
```
are both valid as the doc builder will be able to find them in both cases. I usually prefer the short version for readability. Something you have to use the long version because the object is not exposed at the root of the package. Tough most of the time, it is more for legacy reasons than a proper reason.

**Comment by Wauplin** [comment]
Thanks for the kind words :hugs: I'm always pleased to see high quality PRs to review!

> About enable_webhook and disable_webhook: How do i find out if the server returns something? 😬

Best way to know it is by testing it! :smile: If you run

```py
from huggingface_hub import create_webhook, enable_webhook, disable_webhook

webhook = create_webhook(...) # insert stuff here

# This will raise a Python error if server doesn't return something (at least with the current PR)
disable_webhook(webhook.id)
enable_webhook(webhook.id)
```

^ feel free to create dummy webhooks on your HF repo. You must have a write token (see [authentication](https://huggingface.co/docs/huggingface_hub/quick-start#authentication)).

---
(side note: I'll review all of your PRs on Monday first thing! Had quite a lot on my plate this week :) )

File: src/huggingface_hub/__init__.py
Comment (Wauplin) [inline src/huggingface_hub/__init__.py:153]: `WebhookWatchedItem` should be exported as well here

**Comment by lappemic** [comment]
Hi @Wauplin,

I have included the changes based on your feedback and fixed some issues in the last commits:
1. Adjusted keyword order and making them keyword only
2. Standardized the argument order in all WebhookInfo return values and updated corresponding docstrings, including examples
3. Resolved serialization issue in `create_webhook` and `update_webhook`
4. Fixed JSON parsing issue
5. Added tests

Additionally I verified the return values of `enable_webhook` and `disable_webhook`: Both methods return a value, so I have made no changes made there.

Let me know if this looks good or if i can adjust something. Looking forward to your feedback.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8702]:
```suggestion
        watched_dicts = [asdict(item) if is_dataclass(item) else item for item in watched] if watched is not None else []
```

nit: let's convert from dataclass only if it is a dataclass. This way we allow users to pass a list of dicts directly even if it's not the documented way.

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8684]:
```suggestion
            >>> from huggingface_hub import update_webhook, WebhookWatchedItem
```

File: src/huggingface_hub/hf_api.py
Comment (Wauplin) [inline src/huggingface_hub/hf_api.py:8627]:
```suggestion
        watched_dicts = [asdict(item) if is_dataclass(item) else item for item in watched] if watched is not None else []
```

nit: let's convert from dataclass only if it is a dataclass. This way we allow users to pass a list of dicts directly even if it's not the documented way.

(must add `from dataclasses import is_dataclass` in imports section)

File: tests/test_hf_api.py
Comment (Wauplin) [inline tests/test_hf_api.py:3704]: Really clean and complete tests! Thanks @lappemic!

**Review by Wauplin — COMMENTED** [review]
Great job @lappemic! Everything looks good implementation-wise now :) I left minor comments but that's all.

Last but not least, could you add some documentation to let users know about this new feature? Here is what I would do:
- rename https://huggingface.co/docs/huggingface_hub/guides/webhooks_server to https://huggingface.co/docs/huggingface_hub/guides/webhooks
- add a first section about how to manage webhooks: create, list, enable/disable, delete. Basically the same information / examples as in the docstring but more in the form of a user guide.  
- move all the current content of [Webhooks Server](https://huggingface.co/docs/huggingface_hub/guides/webhooks_server) inside a Webhook Server section in the same page
- add a redirection from https://huggingface.co/docs/huggingface_hub/guides/webhooks_server to https://huggingface.co/docs/huggingface_hub/guides/webhooks => I show you this once the rest is done. It is to avoid breaking previous urls. (see [_redirects.yml](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/_redirects.yml).

In practice it means:
- renaming [webhooks_server.md](https://github.com/huggingface/huggingface_hub/blob/main/docs/source/en/guides/webhooks_server.md) to `webhooks.md`
- updating the [toctree.yml](https://github.com/huggingface/huggingface_hub/blob/2a98fb5a16c6056aae06c22eff81c395eee995a9/docs/source/en/_toctree.yml#L41)
- add content in `webhooks.md`
- add `WebhookInfo` and `WebhookWatchedItem` to the listed dataclasses in the [package reference](https://github.com/huggingface/huggingface_hub/blob/2a98fb5a16c6056aae06c22eff81c395eee995a9/docs/source/en/package_reference/hf_api.md?plain=1#L36)

Please let me know if that's fine for you and if you have any questions. That would make a complete PR for a new feature, from implementation to testing to documenting it. Thanks!

**Comment by lappemic** [comment]
Thanks a lot @Wauplin! 

I would love to finish this off with the added documentation. I ping you asap it's implemented.

**Review by Wauplin — APPROVED** [review]
Thanks a ton for the changes @lappemic! I finally reviewed the last detailed and pushed a few modifications in https://github.com/huggingface/huggingface_hub/pull/2209/commits/e60bbc8224e52f7e136ebeacdb83950c1df97a3b (mostly to fix code styling in CI + switch the 2 sections in the guide). Looks good now so I'll merge it now. Failing tests in the CIs are unrelated to your PR. Thanks again for the hard work! This is very nice!

**Comment by lappemic** [comment]
No worries, well deserved time off i would say! 😄 

Thanks a lot for the guidance and help on this @Wauplin! I really enjoyed the process!

**Pull request merged**
